solutions/haskell/06-set_get/explanation.md

```diff
@@ -0,0 +1,210 @@
+
+We then use `modifyTVar` to change the database by `insert`ing a new or existing `key` and add or replace its `value`, respectively.
+The `insert` function is from the `Data.Map` package because we actually modify the map.
+As per the Redis definition, a successful set command simply returns `"OK"` to the client.
+
+```haskell
+type Command = TVar DB -> IO Response
+
+set :: Key -> Value -> TVar DB -> IO Response
+set key val db = do
+    _ <- atomically $ modifyTVar db $ insert key val
+    return "OK"  -- see the section 3 for an improvement
+```
+
+## 2.2. get
+The parsing of `get` happens in the same way as all other functions.
+You can look at the solution code to see our implementation of `parseGet`.
+
+Similar to set, we have to add the `TVar DB` type to get's signature so that we can read from the database.
+Reading from the database is now such an action as described earlier, where we access the database's current value through the impure `IO` "interface".
+The `readTVarIO` always returns the latest value of our `TVar` database, i.e. the keys and values stored in our map.
+
+Since we are reading from our map, we can use another map function, `findWithDefault`, also from the `Data.Map` package.
+The Redis standard defines that if no key is found, get should return `(nil)`.
+`findWithDefault` accepts a default value in such a case, which we happily use.
+The other two parameters are the `key` and the map, i.e. `out`.
+
+To end up with a concise implementation we can leverage an applicative functor, here.
+Basically, we can apply the `findWithDefault` function to the function that reads the database, `readTVarIO`.
+We could do this by either using `pure (findWithDefault "(nil)" key) <*> readTVarIO db`, or as an alternative, use `findWithDefault "(nil)" key <$> readTVarIO db`, the latter being the infix version of the former.
+
+```haskell
+get :: Key -> TVar DB -> IO Response
+get key db = findWithDefault "(nil)" key <$> readTVarIO db
+```
+
+Such an operation works because `findWithDefault` is an instance of `Functor` which provides the function `fmap` that lets you apply a function to another function.
+You can read more about [the Functor class](https://en.wikibooks.org/wiki/Haskell/The_Functor_class) and [the Applicative class (for applicative functors)](https://en.wikibooks.org/wiki/Haskell/Applicative_functors) following these two links.
+
+# 3. Other improvements
+
+With the `exec` function expecting now a database to be consumed for `set` and `get`, each of the other pattern matches does so, too.
+Therefore, we need to add the `TVar DB` type to `Echo` and `Ping`, even if we do nothing with it.
+Otherwise, we would get an error for having different number of arguments.
+
+By now we are using many constants in the program
+This could get confusing, especially if they are used in multiple different functions.
```

Comment:
I'd agree with this for cases where there is a lot of re-use, but in the context of our solution here I think it makes the program less readable. Most of these constants are only used in 1-2 places - and the places that they are used are functions that specifically deal with the redis spec, so it seems okay to hardcode values there.

Reply:
I'll remove the entire `RedisSpec` and `CmdNames` named fields records, since, to your point, most variables are only used once in the code and limit readability.